src/main/java/org/truffleruby/core/time/RubyDateFormatter.java

```diff
@@ -544,6 +568,168 @@ public static RopeBuilder formatToRopeBuilder(List<Token> compiledPattern, Zoned
+
+                switch (token.getFormat()) {
+                    case FORMAT_ENCODING:
+                    case FORMAT_OUTPUT:
+                        continue;
+
+                    case FORMAT_STRING:
+                        appendRope = token.getRope();
+                        break;
+                    case FORMAT_DAY:
+                        appendRope = RopeConstants.paddedNumber(dt.getDayOfMonth());
+                        break;
+                    case FORMAT_HOUR:
+                        appendRope = RopeConstants.paddedNumber(dt.getHour());
+                        break;
+                    case FORMAT_MINUTES:
+                        appendRope = RopeConstants.paddedNumber(dt.getMinute());
+                        break;
+                    case FORMAT_MONTH:
+                        appendRope = RopeConstants.paddedNumber(dt.getMonthValue());
+                        break;
+                    case FORMAT_SECONDS:
+                        appendRope = RopeConstants.paddedNumber(dt.getSecond());
+                        break;
+
+                    case FORMAT_YEAR_LONG: {
+                        final int value = dt.getYear();
+
+                        if (value < 1000 || value > 9999) {
+                            CompilerDirectives.transferToInterpreterAndInvalidate();
+                            return formatToRopeBuilder(
+                                    compiledPattern,
+                                    dt,
+                                    zone,
+                                    context,
+                                    language,
+                                    currentNode,
+                                    errnoErrorNode).toRope();
+                        }
+
+                        appendRope = new LazyIntRope(value, UTF8Encoding.INSTANCE, 4);
+                    }
+                        break;
+
+                    case FORMAT_NANOSEC: {
+                        final int nano = dt.getNano();
+                        assert nano >= 0;
+                        assert nano < 1000000000;
+
+                        final int digits;
+
+                        if (nano < 10) {
+                            digits = 1;
+                        } else if (nano < 100) {
+                            digits = 2;
+                        } else if (nano < 1000) {
+                            digits = 3;
+                        } else if (nano < 10000) {
+                            digits = 4;
+                        } else if (nano < 100000) {
+                            digits = 5;
+                        } else if (nano < 1000000) {
+                            digits = 6;
+                        } else if (nano < 10000000) {
+                            digits = 7;
+                        } else if (nano < 100000000) {
+                            digits = 8;
+                        } else {
+                            digits = 9;
+                        }
```

Comment:
I feel like _surely_ there's a better way to get the number of digits?

Reply:
Can we simply rely on LazyIntRope computing this? I don't think it's so performance critical here.

Reply:
Could do - this version is simpler as we have more info. Just seemed easy to do?

Reply:
It's really verbose for what it gains, I'd much rather reuse the existing logic.